<a href="https://colab.research.google.com/github/dgullate/Curso-IE/blob/master/week1/pandas_completo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas
![alt text](https://pandas.pydata.org/_static/pandas_logo.png)

Pandas es una librería de python que proporciona unas estructuras de datos fáciles de usar y de alto rendimiento, con objeto de realizar tareas de análisis de datos.

Los objetos principales son los `dataframes`, unas estructuras de datos bidimensionales en forma de tabla

In [0]:
from google.colab import drive
drive.mount('/content/drive')


Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [2]:
print ('I see it in the repo')

I see it in the repo


In [0]:
cd My\ Drive

/content/drive/My Drive


In [0]:
f=open('prueba.txt','w')
f.write('Hola nundo!')
f.close()

In [0]:
!git clone https://github.com/dgullate/Curso-IE.git

Cloning into 'Curso-IE'...
fatal: could not read Username for 'https://github.com': No such device or address


In [0]:
import pandas as pd #Así importamos la librería pandas
import numpy as np

## Creación de dataframes

Podemos crear un dataframe a partir de varias Series: cada una actuará a modo de registro en nuestra tabla. Por ejemplo, crearemos una tabla con tres registros y cuyas columnas sean `Cost`, `Item Purchased` y `Name`:

In [0]:

purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])

Observamos que basta proporcionarle una serie por cada entrada en nuestra tabla. Además, los campos de la Serie deben ser los mismos a lo largo de cada registro, con el fin de que se pueda formar correctamente la tabla.
El argumento index sirve para introducir identificadores a cada registro. Por defecto será un número entero, pero en este caso hemos utilizado strings.

In [0]:
df.head()

Cost Item Purchased   Name
Store 1  22.5       Dog Food  Chris
Store 1   2.5   Kitty Litter  Kevyn
Store 2   5.0      Bird Seed  Vinod

Para seleccionar un subconjunto de registros conociendo su índice, podemos utilizar `loc`

In [0]:
df.loc['Store 1']

Cost Item Purchased   Name
Store 1  22.5       Dog Food  Chris
Store 1   2.5   Kitty Litter  Kevyn

Si además queremos especificar una columna en concreto, podemos decírselo separandolo con una coma

In [0]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

A su vez, estos argumentos pueden ser listas!

In [0]:
df.loc[['Store 1', 'Store 2'], ['Cost','Name']]

Cost   Name
Store 1  22.5  Chris
Store 1   2.5  Kevyn
Store 2   5.0  Vinod

Si en lugar de mediante los strings queremos acceder utilizando las posiciones (como enteros), podemos usar `iloc`.
Por ejemplo, el siguiente comando toma las dos primeras filas y todas las columnas

In [0]:
df.iloc[:3, :]

Cost Item Purchased   Name
Store 1  22.5       Dog Food  Chris
Store 1   2.5   Kitty Litter  Kevyn
Store 2   5.0      Bird Seed  Vinod

<br>
Si queremos borrar algún registro conociendo su índice, basta usar `drop`

In [0]:
df.drop('Store 1')

Cost Item Purchased   Name
Store 2   5.0      Bird Seed  Vinod

In [0]:
df

Cost Item Purchased   Name
Store 1  22.5       Dog Food  Chris
Store 1   2.5   Kitty Litter  Kevyn
Store 2   5.0      Bird Seed  Vinod

Esto es porque `drop` por defecto crea una copia con el nuevo dataframe actualizado. En lugar de `df2 = df.drop(...)` podemos hacer lo siguiente para que la actualización se realice en el mismo lugar de memoria del objeto `df`

In [0]:
df.drop('Store 1', inplace=True)
df

Cost Item Purchased   Name
Store 2   5.0      Bird Seed  Vinod

## Carga de dataframes

En la realidad no trabajaremos con datos directamente creados en el código, si no que éstos ya se encontrarán en archivos del disco duro. Pandas cuenta con funciones para leer el contenido de estos archivos, y volcarlos a un dataframe.

Las funciones que se ocupan de esto comienzan por `pd.read_...`. Existen para documentos excel, html, json. Nosotros utilizaremos la más común, `pd.read_csv`, que lee del respectivo archivo .csv (cada fila es un registro de la tabla, y los valores están separados por comas).

A continuación cargaremos datos `tips.csv` en un `dataframe` de `pandas`. Esta base de datos, contiene información acerca de las cuentas de un restaurante. En concreto contiene las siguientes variables:

* `total_bill`: Valor de la cuenta.
* `tip`: Valor de la propina.
* `sex`: Género del pagador (Femal/Male).
* `smoker`: Variable categórica que indica si el pagador es o no fumador (Yes/No).
* `day`: Día de la semana.
* `time`: Variable categórica que indica si la cuenta corresponde a una comida o una cena (Dinner/Lunch).
* `size`: Número de comensales en la mesa.

In [0]:
url='https://raw.githubusercontent.com/dgullate/Curso-IE/master/Week_1/data/tips.csv?token=Ah4VEo9dUJUyOCk4VzAjbBdTy3fvC3ULks5cqJhCwA%3D%3D'
df = pd.read_csv(url)

In [0]:
df

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
5         25.29  4.71    Male     No   Sun  Dinner     4
6          8.77  2.00    Male     No   Sun  Dinner     2
7         26.88  3.12    Male     No   Sun  Dinner     4
8         15.04  1.96    Male     No   Sun  Dinner     2
9         14.78  3.23    Male     No   Sun  Dinner     2
10        10.27  1.71    Male     No   Sun  Dinner     2
11        35.26  5.00  Female     No   Sun  Dinner     4
12        15.42  1.57    Male     No   Sun  Dinner     2
13        18.43  3.00    Male     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
15        21.58  3.92    Male     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
17        16.29  3.71    Male     No   Sun  Dinner     3
18        16.97  3.50  Female     No   Sun  Dinner     3
19        20.65  3.35    Male     No   Sat  Dinner     3
20        17.92  4.08    Male     No   Sat  Dinner     2
21        20.29  2.75  Female     No   Sat  Dinner     2
22        15.77  2.23  Female     No   Sat  Dinner     2
23        39.42  7.58    Male     No   Sat  Dinner     4
24        19.82  3.18    Male     No   Sat  Dinner     2
25        17.81  2.34    Male     No   Sat  Dinner     4
26        13.37  2.00    Male     No   Sat  Dinner     2
27        12.69  2.00    Male     No   Sat  Dinner     2
28        21.70  4.30    Male     No   Sat  Dinner     2
29        19.65  3.00  Female     No   Sat  Dinner     2
..          ...   ...     ...    ...   ...     ...   ...
214       28.17  6.50  Female    Yes   Sat  Dinner     3
215       12.90  1.10  Female    Yes   Sat  Dinner     2
216       28.15  3.00    Male    Yes   Sat  Dinner     5
217       11.59  1.50    Male    Yes   Sat  Dinner     2
218        7.74  1.44    Male    Yes   Sat  Dinner     2
219       30.14  3.09  Female    Yes   Sat  Dinner     4
220       12.16  2.20    Male    Yes   Fri   Lunch     2
221       13.42  3.48  Female    Yes   Fri   Lunch     2
222        8.58  1.92    Male    Yes   Fri   Lunch     1
223       15.98  3.00  Female     No   Fri   Lunch     3
224       13.42  1.58    Male    Yes   Fri   Lunch     2
225       16.27  2.50  Female    Yes   Fri   Lunch     2
226       10.09  2.00  Female    Yes   Fri   Lunch     2
227       20.45  3.00    Male     No   Sat  Dinner     4
228       13.28  2.72    Male     No   Sat  Dinner     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
230       24.01  2.00    Male    Yes   Sat  Dinner     4
231       15.69  3.00    Male    Yes   Sat  Dinner     3
232       11.61  3.39    Male     No   Sat  Dinner     2
233       10.77  1.47    Male     No   Sat  Dinner     2
234       15.53  3.00    Male    Yes   Sat  Dinner     2
235       10.07  1.25    Male     No   Sat  Dinner     2
236       12.60  1.00    Male    Yes   Sat  Dinner     2
237       32.83  1.17    Male    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

Además del nombre del archivo con los datos, `read_csv` tiene otros argumentos opcionales. Los más utilizados son:
    
* nrows: lee solo cierto número de filas (útil para hacer una prueba antes de procesar un archivo muy grande).
* usecols: el dataframe resultante solo tendrá estas columnas
* dtype: especifica el tipo de cada columna (por defecto, pandas trata de inferirlo automáticamente de los datos, pero a veces es necesario: por ejemplo para que lea un número como string en lugar de int).

## Análisis exploratorio

Una vez ya tenemos los datos cargados, podemos echarles un vistazo mediante las siguientes cuatro funciones: `head` y `tail` para observar las primeras y últimas filas, respectivamente, `describe` para obtener algunos estadísticos básicos, e `info` para obtener datos sobre los tipos

In [0]:
df.head(10) #Entre paréntesis, el número de entradas que queremos mostrar. Por defecto,
            # este número es 5.

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
6        8.77  2.00    Male     No  Sun  Dinner     2
7       26.88  3.12    Male     No  Sun  Dinner     4
8       15.04  1.96    Male     No  Sun  Dinner     2
9       14.78  3.23    Male     No  Sun  Dinner     2

In [0]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [0]:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

Para datos de tipo numérico, `describe` proporciona información del número de valores no omitidos (count), la media, la desviación estándar, los valores mínimo y máximo y algunos cuantiles, entre ellos la mediana (50%).

En el caso de que la tabla contenga tanto datos numéricos como categóricos, `describe` por defecto solo muestra los estadísticos para las numéricas. Si queremos que también muestre los otros:

In [0]:
df.describe(include='all')

total_bill         tip   sex smoker  day    time        size
count   244.000000  244.000000   244    244  244     244  244.000000
unique         NaN         NaN     2      2    4       2         NaN
top            NaN         NaN  Male     No  Sat  Dinner         NaN
freq           NaN         NaN   157    151   87     176         NaN
mean     19.785943    2.998279   NaN    NaN  NaN     NaN    2.569672
std       8.902412    1.383638   NaN    NaN  NaN     NaN    0.951100
min       3.070000    1.000000   NaN    NaN  NaN     NaN    1.000000
25%      13.347500    2.000000   NaN    NaN  NaN     NaN    2.000000
50%      17.795000    2.900000   NaN    NaN  NaN     NaN    2.000000
75%      24.127500    3.562500   NaN    NaN  NaN     NaN    3.000000
max      50.810000   10.000000   NaN    NaN  NaN     NaN    6.000000

Para las variables categóricas, se muestran los nuevos estadísticos de: número de valores únicos, valor más frecuente (top) y frecuencia de éste. Por ejemplo, para el caso del sexo, hay 244 anotaciones, y el valor más frecuente es Male (157 de 244), además de tener dos valores únicos (por lo que asumiremos que el otro valor es Female)

In [0]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
total_bill    244 non-null float64
tip           244 non-null float64
sex           244 non-null object
smoker        244 non-null object
day           244 non-null object
time          244 non-null object
size          244 non-null int64
dtypes: float64(2), int64(1), object(4)
memory usage: 13.4+ KB


## Consultas de dataframes

Hasta ahora hemos visto como acceder a ciertos registros y/o ciertas columnas de nuestra tabla. Veremos ahora algunas funciones nuevas para realizar consultas más dinámicas.

Una forma rápida de acceder a una columna entera es

In [0]:
df['tip'].head()

0    1.01
1    1.66
2    3.50
3    3.31
4    3.61
Name: tip, dtype: float64

También se podría hacer

In [0]:
df.tip.head()

0    1.01
1    1.66
2    3.50
3    3.31
4    3.61
Name: tip, dtype: float64

Como pandas va sobre numpy, podemos utilizar una sintaxis muy similar

In [0]:
df['size'].median(), df['size'].mean()

(2.0, 2.569672131147541)

In [0]:
df['size'] > df['size'].median()

Como lo anterior nos ha generado un array booleano, podemos utilizarlo para seleccionar filas en nuestro dataframe!

In [0]:
mesas_grandes = df[ df['size'] > df['size'].median() ]
mesas_grandes.head()

total_bill   tip     sex smoker  day    time  size
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
7       26.88  3.12    Male     No  Sun  Dinner     4

Observamos que conserva los índices de la tabla original. Si queremos resetearlos, basta con hacer

In [0]:
mesas_grandes.reset_index(inplace=True, drop=True)
mesas_grandes.head()

total_bill   tip     sex smoker  day    time  size
0       10.34  1.66    Male     No  Sun  Dinner     3
1       21.01  3.50    Male     No  Sun  Dinner     3
2       24.59  3.61  Female     No  Sun  Dinner     4
3       25.29  4.71    Male     No  Sun  Dinner     4
4       26.88  3.12    Male     No  Sun  Dinner     4

Los criterios de filtrado de filas pueden ser tan complejos como queramos, utilizando los operadores lógicos

In [0]:
df[ (df['size'] > df['size'].median()) & (df['sex'] == 'Female') ].head()

total_bill   tip     sex smoker  day    time  size
4        24.59  3.61  Female     No  Sun  Dinner     4
11       35.26  5.00  Female     No  Sun  Dinner     4
16       10.33  1.67  Female     No  Sun  Dinner     3
18       16.97  3.50  Female     No  Sun  Dinner     3
33       20.69  2.45  Female     No  Sat  Dinner     4


##    Ejercicio 1
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 


Carga la tabla `olympics.csv`. Si observas el archivo, hay que saltarse la primera fila (skiprows). Además, indica a pandas que la primera columna sea el index (index_col=0). Realiza un análisis exploratorio de los datos

In [0]:
url='https://raw.githubusercontent.com/dgullate/Curso-IE/master/Week_1/data/olympics.csv?token=Ah4VEg7qZlveLgZOUytJTsWkuO_vjc69ks5cqJnhwA%3D%3D'
olympics = pd.read_csv(url, skiprows=1, index_col=0)

Para ayudarnos en la exploración, ejecuta el siguiente código para renombrar las columnas

In [0]:
for col in olympics.columns:
    if col[:2]=='01':
        olympics.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        olympics.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        olympics.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        olympics.rename(columns={col:'#' + col[1:]}, inplace=True) 
        
names_ids = olympics.index.str.split('\s\(') # split the index by '('

olympics.index = names_ids.str[0] # the [0] element is the country name (new index) 
olympics['ID'] = names_ids.str[1].str[:3] # the [1] element is the abbreviation or ID (take first 3 characters from that)
        
olympics =  olympics.drop('Totals')
olympics.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

In [0]:
olympics.describe()

# Summer        Gold      Silver      Bronze        Total  \
count  146.000000  146.000000  146.000000  146.000000   146.000000   
mean    13.383562   32.938356   32.705479   35.136986   100.780822   
std      7.006657   96.745184   80.648421   77.118416   252.700436   
min      1.000000    0.000000    0.000000    0.000000     0.000000   
25%      8.000000    0.000000    1.000000    1.000000     2.000000   
50%     13.000000    3.000000    4.000000    6.000000    12.000000   
75%     18.000000   23.000000   26.750000   28.750000    85.250000   
max     27.000000  976.000000  757.000000  666.000000  2399.000000   

         # Winter      Gold.1    Silver.1    Bronze.1     Total.1     # Games  \
count  146.000000  146.000000  146.000000  146.000000  146.000000  146.000000   
mean     6.595890    6.568493    6.561644    6.493151   19.623288   19.979452   
std      7.349012   18.975086   18.606996   17.499654   54.640914   13.084062   
min      0.000000    0.000000    0.000000    0.000000    0.000000    1.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000   11.000000   
50%      5.000000    0.000000    0.000000    0.000000    0.000000   15.000000   
75%     10.000000    0.750000    1.750000    1.000000    4.750000   26.000000   
max     22.000000  118.000000  111.000000  100.000000  329.000000   49.000000   

            Gold.2    Silver.2    Bronze.2  Combined total  
count   146.000000  146.000000  146.000000      146.000000  
mean     39.506849   39.267123   41.630137      120.404110  
std     109.932471   93.984393   89.826117      291.988913  
min       0.000000    0.000000    0.000000        1.000000  
25%       0.000000    1.000000    1.000000        2.250000  
50%       3.000000    4.000000    6.500000       12.000000  
75%      24.500000   28.000000   29.000000       87.750000  
max    1072.000000  859.000000  750.000000     2681.000000

¿Qué país ha obtenido más medallas de oro en juegos de verano?

In [0]:
olympics[olympics['Gold']==max(olympics['Gold'])]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
United States        26   976     757     666   2399        22      96   

               Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
United States       102        84      282       48    1072       859   

               Bronze.2  Combined total   ID  
United States       750            2681  USA

In [0]:
olympics.Gold.idxmax()

'United States'

Obtén los países que hayan obtenido más medallas (cuenta cada tipo de medalla por igual) en inverno que en verano 

In [0]:
list(olympics[olympics['Total'] < olympics['Total.1']].index)

['Austria', 'Liechtenstein', 'Norway']

## Creando nuevas columnas y aplicando funciones sobre columnas

Si queremos añadir una nueva columna a nuestra tabla, es tan sencillo como realizar una nueva asignación

In [0]:
df['new_col'] = None
df.head()

También podemos crear columnas utilizando funciones de numpy, por ejemplo, ahora la rellenaremos con valores aleatorios provenientes de una distribución normal

In [0]:
df['new_col'] = np.random.randn(len(df))
df.head()

total_bill   tip     sex smoker  day    time  size   new_col
0       16.99  1.01  Female     No  Sun  Dinner     2  1.015183
1       10.34  1.66    Male     No  Sun  Dinner     3  1.252734
2       21.01  3.50    Male     No  Sun  Dinner     3  1.575078
3       23.68  3.31    Male     No  Sun  Dinner     2  0.221623
4       24.59  3.61  Female     No  Sun  Dinner     4  0.039259

También es posible combinarlos con otras columnas de la tabla, siempre que las dimensiones lo permitan (en este caso, que tengan el mismo número de elementos)

In [0]:
del df['new_col']
df['total_bill_rand'] = np.random.randn(len(df)) + df['total_bill']
df.head()

total_bill   tip     sex smoker  day    time  size  total_bill_rand
0       16.99  1.01  Female     No  Sun  Dinner     2        18.200907
1       10.34  1.66    Male     No  Sun  Dinner     3        10.706272
2       21.01  3.50    Male     No  Sun  Dinner     3        21.085219
3       23.68  3.31    Male     No  Sun  Dinner     2        24.703094
4       24.59  3.61  Female     No  Sun  Dinner     4        26.438620

Una función importante de pandas es `apply`, que permite aplicar cualquier función (de numpy, pandas, o definida por nosotros) a lo largo de varias columnas.

Por ejemplo, supongamos que queremos calcular la raíz cuadrada a las columnas total_bill, tip y total_bill_rand:

In [0]:
df[['total_bill', 'tip', 'total_bill_rand']].apply(np.sqrt).head()

total_bill       tip  total_bill_rand
0    4.121893  1.004988         4.266252
1    3.215587  1.288410         3.272044
2    4.583667  1.870829         4.591864
3    4.866210  1.819341         4.970221
4    4.958831  1.900000         5.141850

In [0]:
df.apply(np.max)

total_bill      50.81
tip                10
sex              Male
smoker            Yes
day              Thur
time            Lunch
size                6
new_col       2.87327
dtype: object

Ahora veremos otro ejemplo con una función creada por nosotros:

In [0]:
def filtra_peq(x, y):  # x es la columna de valores
    x[x<y] = 0
    return x

In [0]:
df[['total_bill']].apply(filtra_peq, args=[df['total_bill'].mean()]).head()

total_bill
0        0.00
1        0.00
2       21.01
3       23.68
4       24.59

Otra forma de hacerlo, con `where`: si la condición es cierta, deja el valor igual, pero si es falsa, reemplaza con el segundo argumento. 

In [0]:
mask=df['total_bill']>df['total_bill'].mean()
df['total_bill'].where(mask,0).head()

0     0.00
1     0.00
2    21.01
3    23.68
4    24.59
Name: total_bill, dtype: float64

### Ejercicio

Trabajaremos sobre olympics. Obtén los países que hayan obtenido más puntos en inverno que en verano. Se asignan puntos de la siguiente forma: bronce 1, plata 2, oro 3.

In [0]:
olympics['points_summer'] = 3*olympics['Gold'] + 2*olympics['Silver'] + olympics['Bronze']
olympics['points_winter'] = 3*olympics['Gold.1'] + 2*olympics['Silver.1'] + olympics['Bronze.1']

In [0]:
olympics[olympics['points_summer'] < olympics['points_winter']]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Austria              26    18      33      35     86        22      59   
Liechtenstein        16     0       0       0      0        18       2   
Norway               24    56      49      43    148        22     118   

               Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Austria              78        81      218       48      77       111   
Liechtenstein         2         5        9       34       2         2   
Norway              111       100      329       46     174       160   

               Bronze.2  Combined total   ID  points_summer  points_winter  
Austria             116             304  AUT            155            414  
Liechtenstein         5               9  LIE              0             15  
Norway              143             477  NOR            309            676

¿Qué país tiene la menor diferencia relativa de puntos entre verano e invierno? Es decir, para cada país, calcula
$$
dif\_rel = \frac{|points\_ summer - points\_winter|}{points\_summer + points\_winter} 
$$

In [0]:
import numpy as np
olympics['dif_rel'] = np.abs(olympics['points_summer'] - olympics['points_winter'])/(olympics['points_summer'] + olympics['points_winter'])

In [0]:
olympics[olympics['dif_rel'] == min(olympics['dif_rel'])]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
Luxembourg        22     1       1       0      2         8       0         2   

            Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Luxembourg         0        2       30       1         3         0   

            Combined total   ID  points_summer  points_winter   dif_rel  
Luxembourg               4  LUX              5              4  0.111111

## Fusionando dataframes

Es habitual tener la información repartida en dos o más tablas. Para ello, pandas cuenta con la operación `merge`, que permite fusionar registros provenientes de diferentes tablas (similar al join de SQL). Empezamos con unos datos de ejemplo:

In [0]:
staff_df = pd.DataFrame([{'Nombre': 'Kelly', 'Rol': 'Director de RRHH'},
                         {'Nombre': 'Sally', 'Rol': 'Profesor'},
                         {'Nombre': 'James', 'Rol': 'Secretario'}])
staff_df = staff_df.set_index('Nombre')
school_df = pd.DataFrame([{'Nombre': 'James', 'Facultad': 'Negocios'},
                           {'Nombre': 'Mike', 'Facultad': 'Derecho'},
                           {'Nombre': 'Sally', 'Facultad': 'Ingeniería'}])
school_df = school_df.set_index('Nombre')
print(staff_df.head())
print()
print(school_df.head())

                     Rol
Nombre                  
Kelly   Director de RRHH
Sally           Profesor
James         Secretario

          Facultad
Nombre            
James     Negocios
Mike       Derecho
Sally   Ingeniería


`merge` recibe dos tablas, y el argumento `how`, y una columna en cada tabla que servirá para establecer la correspondencia (en este caso, tomará la columna índice de cada tabla).

Veremos todas las posibilidades de este argumento para entender las diferencias.

Con `outer` la tabla combinada presenta los registros de la tabla izquierda (`staff_df`) y de la tabla derecha (`school_df`) aunque no tengan correspondencia. Por ejemplo, vemos que James era Secretario en la tabla izquierda y pertenecía a la Facultad de Negocios, así que la tabla resultante contiene esa observación. Sin embargo, Kelly sólo figuraba en la tabla Rol, por tanto aparece que su Facultad es Nan:

In [0]:
pd.merge(staff_df, school_df, how='outer', left_index=True, right_index=True)

Rol    Facultad
Nombre                              
James         Secretario    Negocios
Kelly   Director de RRHH         NaN
Mike                 NaN     Derecho
Sally           Profesor  Ingeniería

In [0]:
pd.merge(staff_df, school_df, how='inner', left_index=True, right_index=True)

Rol    Facultad
Nombre                        
Sally     Profesor  Ingeniería
James   Secretario    Negocios

`left` solo tiene en cuenta aquellas claves (en este caso el Nombre) que figuren en la tabla de la izquierda:

In [0]:
pd.merge(staff_df, school_df, how='left', left_index=True, right_index=True)

Rol    Facultad
Nombre                              
Kelly   Director de RRHH         NaN
Sally           Profesor  Ingeniería
James         Secretario    Negocios

`right` es análogo a `left`:

In [0]:
pd.merge(staff_df, school_df, how='right', left_index=True, right_index=True)

Rol    Facultad
Nombre                        
James   Secretario    Negocios
Mike           NaN     Derecho
Sally     Profesor  Ingeniería

Ahora, asignaremos nuevos índices numéricos, y realizaremos un `merge`, especificando qué columnas queremos utilizar como claves en cada tabla

In [0]:
staff_df = staff_df.reset_index()
school_df = school_df.reset_index()
pd.merge(staff_df, school_df, how='left', left_on='Nombre', right_on='Nombre')

Nombre               Rol    Facultad
0  Kelly  Director de RRHH         NaN
1  Sally          Profesor  Ingeniería
2  James        Secretario    Negocios

Una observación: si además de la clave, en cada tabla tenemos columnas que se llaman igual (Ubicación en el ejemplo siguiente), al hacer el `merge` pandas añadirá sufijos para evitar la colisión de nombres

In [0]:
staff_df = pd.DataFrame([{'Nombre': 'Kelly', 'Rol': 'Director de RRHH', 'Ubicación': 'State Street'},
                         {'Nombre': 'Sally', 'Rol': 'Profesor', 'Ubicación': 'Washington Avenue'},
                         {'Nombre': 'James', 'Rol': 'Secretario', 'Ubicación': 'Washington Avenue'}])
school_df = pd.DataFrame([{'Nombre': 'James', 'Facultad': 'Negocios', 'Ubicación': '1024 Billiard Avenue'},
                           {'Nombre': 'Mike', 'Facultad': 'Derecho', 'Ubicación': 'Fraternity House #22'},
                           {'Nombre': 'Sally', 'Facultad': 'Ingeniería', 'Ubicación': '512 Wilson Crescent'}])
print(staff_df)
print()
print(school_df)

  Nombre               Rol          Ubicación
0  Kelly  Director de RRHH       State Street
1  Sally          Profesor  Washington Avenue
2  James        Secretario  Washington Avenue

     Facultad Nombre             Ubicación
0    Negocios  James  1024 Billiard Avenue
1     Derecho   Mike  Fraternity House #22
2  Ingeniería  Sally   512 Wilson Crescent


In [0]:
pd.merge(staff_df, school_df, how='left', left_on='Nombre', right_on='Nombre')

Nombre               Rol        Ubicación_x    Facultad  \
0  Kelly  Director de RRHH       State Street         NaN   
1  Sally          Profesor  Washington Avenue  Ingeniería   
2  James        Secretario  Washington Avenue    Negocios   

            Ubicación_y  
0                   NaN  
1   512 Wilson Crescent  
2  1024 Billiard Avenue

También es posible utilizar como clave varias columnas:

In [0]:
staff_df = pd.DataFrame([{'Nombre': 'Sally', 'Apellido': 'Desjardins', 'Rol': 'Director de RRHH'},
                         {'Nombre': 'Sally', 'Apellido': 'Brooks', 'Rol': 'Profesor'},
                         {'Nombre': 'James', 'Apellido': 'Wilde', 'Rol': 'Secretario'}])
school_df = pd.DataFrame([{'Nombre': 'James', 'Apellido': 'Hammond', 'Facultad': 'Negocios'},
                           {'Nombre': 'Sally', 'Apellido': 'Smith', 'Facultad': 'Derecho'},
                           {'Nombre': 'Sally', 'Apellido': 'Brooks', 'Facultad': 'Ingeniería'}])


print(staff_df)
print()
print(school_df)


     Apellido Nombre               Rol
0  Desjardins  Sally  Director de RRHH
1      Brooks  Sally          Profesor
2       Wilde  James        Secretario

  Apellido    Facultad Nombre
0  Hammond    Negocios  James
1    Smith     Derecho  Sally
2   Brooks  Ingeniería  Sally


In [0]:
pd.merge(staff_df, school_df, how='inner', left_on=['Nombre','Apellido'], right_on=['Nombre','Apellido'])

Apellido Nombre       Rol    Facultad
0   Brooks  Sally  Profesor  Ingeniería

## Ejercicio 2 
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 

Seguimos con las Olimpiadas. Carga la tabla `population.csv`, que para cada país nos da su población. Para simplificar, nos quedaremos con las poblaciones del año 2016 (haz un filtro). Después, combina las tablas olympics y population (de forma que no queden valores missing) para que en cada registro tengamos al menos el nombre del país, el número de medallas de oro, y su población. 

¿Qué país es el que tiene mayor ratio de medallas de oro por habitante?

In [0]:
url='https://raw.githubusercontent.com/dgullate/Curso-IE/master/Week_1/data/population.csv?token=Ah4VEg_P7qLkucZLQcXcgCx5-rP56fXdks5cqKW_wA%3D%3D'
pop = pd.read_csv(url)

In [0]:
pop.tail()

Country Name Country Code  Year       Value
14880     Zimbabwe          ZWE  2012  14710826.0
14881     Zimbabwe          ZWE  2013  15054506.0
14882     Zimbabwe          ZWE  2014  15411675.0
14883     Zimbabwe          ZWE  2015  15777451.0
14884     Zimbabwe          ZWE  2016  16150362.0

In [0]:
pop = pop[pop['Year'] == 2016]

In [0]:
pop.reset_index()

In [0]:
olympics = olympics.reset_index()

In [0]:
oly_pop = pd.merge(olympics, pop, how='inner', left_on='index', right_on='Country Name')
oly_pop.head()

index  # Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
0  Afghanistan        13     0       0       2      2         0       0   
1      Algeria        12     5       2       8     15         3       0   
2    Argentina        23    18      24      28     70        18       0   
3      Armenia         5     1       2       9     12         6       0   
4    Australia        25   139     152     177    468        18       5   

   Silver.1  Bronze.1     ...      # Games  Gold.2  Silver.2  Bronze.2  \
0         0         0     ...           13       0         0         2   
1         0         0     ...           15       5         2         8   
2         0         0     ...           41      18        24        28   
3         0         0     ...           11       1         2         9   
4         3         4     ...           43     144       155       181   

   Combined total   ID Country Name Country Code  Year       Value  
0               2  AFG  Afghanistan          AFG  2016  34656032.0  
1              15  ALG      Algeria          DZA  2016  40606052.0  
2              70  ARG    Argentina          ARG  2016  43847430.0  
3              12  ARM      Armenia          ARM  2016   2924816.0  
4             480  AUS    Australia          AUS  2016  24127159.0  

[5 rows x 21 columns]

In [0]:
oly_pop['ratio'] = oly_pop['Gold'] / oly_pop['Value']

In [0]:
oly_pop[['Country Name', 'ratio']].sort_values(by='ratio', ascending=False).reset_index(drop=True).head(10)

Country Name     ratio
0      Finland  0.000018
1      Hungary  0.000017
2       Sweden  0.000014
3       Norway  0.000011
4      Grenada  0.000009
5  New Zealand  0.000009
6      Denmark  0.000008
7     Bulgaria  0.000007
8      Estonia  0.000007
9         Cuba  0.000006

Now do the same thing as before, but rank the 10 countries that have the largest ratio of gold medals to GDP. Can you guess before which will be the first ?

In [0]:
url='https://raw.githubusercontent.com/dgullate/Curso-IE/master/Week_1/data/gdp_csv.csv?token=Ah4VEh7CPzA7lzVfizWSEaX-V_TMFxt3ks5cqKk3wA%3D%3D'
gdp = pd.read_csv(url)

In [0]:
gdp = gdp[gdp['Year'] == 2015]
gdp_oly_pop= pd.merge(oly_pop, gdp, how='inner', left_on='index', right_on='Country Name')
gdp_oly_pop['gold2gdp_ratio']=gdp_oly_pop['Gold']/gdp_oly_pop['Value_y']
gdp_oly_pop[['Country Name_x', 'gold2gdp_ratio']].sort_values(by='gold2gdp_ratio', ascending=False).reset_index(drop=True).head(10)

Country Name_x  gold2gdp_ratio
0        Hungary    1.359060e-09
1        Jamaica    1.198290e-09
2       Bulgaria    1.015954e-09
3        Grenada    1.003001e-09
4           Cuba    8.263249e-10
5        Romania    4.946291e-10
6        Finland    4.345220e-10
7        Georgia    4.287691e-10
8        Estonia    3.988132e-10
9          Kenya    3.920490e-10

## Agrupando valores en dataframes

Una función muy importante de pandas es `groupby`, con  la que podremos realizar el siguiente proceso:

* Separar la tabla en grupos en función de algún criterio
* Aplicar alguna operación a cada grupo de forma independiente
* Combinar los resultados en una nueva tabla

Empezamos con la creación de grupos. El criterio de separación más habitual es mediante el valor de alguna variable categórica, por ejemplo el sexo, el día de la semana, etc...

Vamos a hacer algunos ejemplos sobre la tabla `tips.csv`

In [0]:
for name, group in df.groupby(by=['sex']):
    print(name)
    print(group)

Female
     total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
11        35.26  5.00  Female     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
18        16.97  3.50  Female     No   Sun  Dinner     3
21        20.29  2.75  Female     No   Sat  Dinner     2
22        15.77  2.23  Female     No   Sat  Dinner     2
29        19.65  3.00  Female     No   Sat  Dinner     2
32        15.06  3.00  Female     No   Sat  Dinner     2
33        20.69  2.45  Female     No   Sat  Dinner     4
37        16.93  3.07  Female     No   Sat  Dinner     3
51        10.29  2.60  Female     No   Sun  Dinner     2
52        34.81  5.20  Female     No   Sun  Dinner     4
57        26.41  1.50  Female     No   Sat  Dinner     2
66        16.45  2.47  Female     No   Sat  Dinner     2
67         3.07  1.00  F

También podemos especificar varias variables para agrupar. En este caso como hay dos valores por sexo, y seis por tamaño de mesa, se crearán 12 grupos

In [0]:
for name, group in df.groupby(by=['sex', 'size']):
    print(name)
    print(group)

('Female', 1)
     total_bill   tip     sex smoker   day    time  size
67         3.07  1.00  Female    Yes   Sat  Dinner     1
82        10.07  1.83  Female     No  Thur   Lunch     1
111        7.25  1.00  Female     No   Sat  Dinner     1
('Female', 2)
     total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
14        14.83  3.02  Female     No   Sun  Dinner     2
21        20.29  2.75  Female     No   Sat  Dinner     2
22        15.77  2.23  Female     No   Sat  Dinner     2
29        19.65  3.00  Female     No   Sat  Dinner     2
32        15.06  3.00  Female     No   Sat  Dinner     2
51        10.29  2.60  Female     No   Sun  Dinner     2
57        26.41  1.50  Female     No   Sat  Dinner     2
66        16.45  2.47  Female     No   Sat  Dinner     2
72        26.86  3.14  Female    Yes   Sat  Dinner     2
73        25.28  5.00  Female    Yes   Sat  Dinner     2
74        14.73  2.20  Female     No   Sat  Dinner     2
92 

Una vez, hemos realizado los grupos, podemos aplicar funciones sobre ellos. Por ejemplo, vamos a calcular la media de la factura para cada grupo

In [0]:
for name, group in df.groupby(by=['sex', 'size']):
    print(name, group['total_bill'].mean())
    # equivalente a np.mean(group['total_bill'])

('Female', 1) 6.796666666666667
('Female', 2) 15.339482758620687
('Female', 3) 21.54714285714286
('Female', 4) 30.278888888888886
('Female', 5) 29.85
('Female', 6) 28.425
('Male', 1) 8.58
('Male', 2) 17.10408163265306
('Male', 3) 24.28708333333333
('Male', 4) 28.07821428571429
('Male', 5) 30.122500000000002
('Male', 6) 41.235


No obstante, la formulación anterior puede no resultar cómoda a la hora de crear una nueva tabla con los resultados. Por eso, pandas incorpora la función `agg`:

In [0]:
df_bill = df.groupby(['sex']).agg({'total_bill': np.mean})
df_bill

total_bill
sex               
Female   18.056897
Male     20.744076

In [0]:
df_bill = df.groupby(['sex', 'size']).agg({'total_bill': np.mean})
df_bill

total_bill
sex    size            
Female 1       6.796667
       2      15.339483
       3      21.547143
       4      30.278889
       5      29.850000
       6      28.425000
Male   1       8.580000
       2      17.104082
       3      24.287083
       4      28.078214
       5      30.122500
       6      41.235000

Utilizando `pivot` podemos pasar la tabla a una vista matricial para ver mejor los resultados

In [0]:
df_bill.reset_index().pivot(index='sex', columns='size')

total_bill                                                  
size            1          2          3          4        5       6
sex                                                                
Female   6.796667  15.339483  21.547143  30.278889  29.8500  28.425
Male     8.580000  17.104082  24.287083  28.078214  30.1225  41.235

A `agg` podemos pasarle más argumentos si queremos calcular más columnas. Con esta otra sintaxis, agg aplicará las funciones mean y std a todo el dataframe anterior (hemos seleccionado las columnas total_bill y tip previamente)

In [0]:
df_bill_tips = df.groupby(['sex'])['total_bill', 'tip'].agg([np.mean, np.std])

In [0]:
df_bill_tips

total_bill                 tip          
             mean       std      mean       std
sex                                            
Female  18.056897  8.009209  2.833448  1.159495
Male    20.744076  9.246469  3.089618  1.489102

## Ejercicio 3
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 

Para la tabla de tips, selcciona aquellas observaciones cuyas `total_bill` sean mayores que las respectivas medias + 1 desviación estándar de su día de la semana.

Indicación: crea primero una tabla pequeña donde para cada día de la semana aparezca la media, y luego haz un `merge`
 de esta tabla con la original, para que cada observación tenga anotada su correspondiente media + std.

In [0]:
df_tips = df.groupby(['day'])['total_bill'].agg([np.mean, np.std])

In [0]:
df_tips = df_tips.reset_index()


In [0]:
df_tips

day       mean       std
0   Fri  17.151579  8.302660
1   Sat  20.441379  9.480419
2   Sun  21.410000  8.832122
3  Thur  17.682742  7.886170

In [0]:
df_augmented= pd.merge(df, df_tips, how='inner', left_on='day', right_on='day')
df_augmented.head()

total_bill   tip     sex smoker  day    time  size   mean       std
0       16.99  1.01  Female     No  Sun  Dinner     2  21.41  8.832122
1       10.34  1.66    Male     No  Sun  Dinner     3  21.41  8.832122
2       21.01  3.50    Male     No  Sun  Dinner     3  21.41  8.832122
3       23.68  3.31    Male     No  Sun  Dinner     2  21.41  8.832122
4       24.59  3.61  Female     No  Sun  Dinner     4  21.41  8.832122

In [0]:
df_augmented[df_augmented['total_bill'] > df_augmented['mean'] + df_augmented['std']]

total_bill    tip     sex smoker   day    time  size       mean       std
11        35.26   5.00  Female     No   Sun  Dinner     4  21.410000  8.832122
22        30.40   5.60    Male     No   Sun  Dinner     4  21.410000  8.832122
25        32.40   6.00    Male     No   Sun  Dinner     4  21.410000  8.832122
30        34.81   5.20  Female     No   Sun  Dinner     4  21.410000  8.832122
34        38.07   4.00    Male     No   Sun  Dinner     3  21.410000  8.832122
45        48.17   5.00    Male     No   Sun  Dinner     6  21.410000  8.832122
56        31.71   4.50    Male     No   Sun  Dinner     4  21.410000  8.832122
58        31.85   3.18    Male    Yes   Sun  Dinner     2  21.410000  8.832122
60        32.90   3.11    Male    Yes   Sun  Dinner     2  21.410000  8.832122
64        34.63   3.55    Male    Yes   Sun  Dinner     2  21.410000  8.832122
65        34.65   3.68    Male    Yes   Sun  Dinner     4  21.410000  8.832122
67        45.35   3.50    Male    Yes   Sun  Dinner     3  21.410000  8.832122
69        40.55   3.00    Male    Yes   Sun  Dinner     2  21.410000  8.832122
72        30.46   2.00    Male    Yes   Sun  Dinner     5  21.410000  8.832122
80        39.42   7.58    Male     No   Sat  Dinner     4  20.441379  9.480419
96        31.27   5.00    Male     No   Sat  Dinner     3  20.441379  9.480419
98        38.01   3.00    Male    Yes   Sat  Dinner     4  20.441379  9.480419
101       48.27   6.73    Male     No   Sat  Dinner     4  20.441379  9.480419
119       44.30   2.50  Female    Yes   Sat  Dinner     3  20.441379  9.480419
131       50.81  10.00    Male    Yes   Sat  Dinner     3  20.441379  9.480419
134       38.73   3.00    Male    Yes   Sat  Dinner     4  20.441379  9.480419
137       30.06   2.00    Male    Yes   Sat  Dinner     3  20.441379  9.480419
139       48.33   9.00    Male     No   Sat  Dinner     4  20.441379  9.480419
146       30.14   3.09  Female    Yes   Sat  Dinner     4  20.441379  9.480419
157       32.83   1.17    Male    Yes   Sat  Dinner     2  20.441379  9.480419
158       35.83   4.67  Female     No   Sat  Dinner     3  20.441379  9.480419
163       27.20   4.00    Male     No  Thur   Lunch     4  17.682742  7.886170
169       32.68   5.00    Male    Yes  Thur   Lunch     2  17.682742  7.886170
171       34.83   5.17  Female     No  Thur   Lunch     4  17.682742  7.886170
184       29.80   4.20  Female     No  Thur   Lunch     6  17.682742  7.886170
200       34.30   6.70    Male     No  Thur   Lunch     6  17.682742  7.886170
201       41.19   5.00    Male     No  Thur   Lunch     5  17.682742  7.886170
202       27.05   5.00  Female     No  Thur   Lunch     6  17.682742  7.886170
210       28.44   2.56    Male    Yes  Thur   Lunch     2  17.682742  7.886170
215       43.11   5.00  Female    Yes  Thur   Lunch     4  17.682742  7.886170
225       28.97   3.00    Male    Yes   Fri  Dinner     2  17.151579  8.302660
230       40.17   4.73    Male    Yes   Fri  Dinner     4  17.151579  8.302660
231       27.28   4.00    Male    Yes   Fri  Dinner     2  17.151579  8.302660

## Ejercicio 4
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 

La base de datos `income` se utiliza para predecir si el salario de una persona es mayor o menor de 50.000 dolares anuales. Para cada persona, contiene la siguiente información:
* age: La edad de la persona en cuestión.
* education: nivel de estudios.
* marital.status: estado civil.
* relationship: puesto que ocupa en su familia.
* race: raza.
* sex: género.
* hourspeerweek: horas de trabajo por semana.
* nativecountry: país de origen.
* income: variable que indica si el salario es mayor o menor de 50k.

Lee los datos, y elimina cualquier fila que contenga algún valor missing. Puedes utilizar en método `dropna`. También, por comodidad, sustituye los valores de `Income` por 0 (si el salario es menor o igual a 50k) y 1 (si es mayor).

In [0]:
url='https://raw.githubusercontent.com/dgullate/Curso-IE/master/Week_1/data/income.csv?token=Ah4VEhWjRnc0HbDfNneCcguB6hDE7hsrks5cqN84wA%3D%3D'
inc = pd.read_csv(url)

In [0]:
inc.head()

age  education      marital.status   relationship   race     sex  \
0   39  Bachelors       Never-married  Not-in-family  White    Male   
1   50  Bachelors  Married-civ-spouse        Husband  White    Male   
2   38    HS-grad            Divorced  Not-in-family  White    Male   
3   53       11th  Married-civ-spouse        Husband  Black    Male   
4   28  Bachelors  Married-civ-spouse           Wife  Black  Female   

   hoursperweek  nativecountry income  
0            40  United-States  <=50K  
1            13  United-States  <=50K  
2            40  United-States  <=50K  
3            40  United-States  <=50K  
4            40           Cuba  <=50K

In [0]:
inc.dropna(inplace=True)
inc.loc[inc.income == '<=50K', "income"] = 0
inc.loc[inc.income == '>50K', "income"] = 1
inc.head(20)

age     education         marital.status   relationship  \
0    39     Bachelors          Never-married  Not-in-family   
1    50     Bachelors     Married-civ-spouse        Husband   
2    38       HS-grad               Divorced  Not-in-family   
3    53          11th     Married-civ-spouse        Husband   
4    28     Bachelors     Married-civ-spouse           Wife   
5    37       Masters     Married-civ-spouse           Wife   
6    49           9th  Married-spouse-absent  Not-in-family   
7    52       HS-grad     Married-civ-spouse        Husband   
8    31       Masters          Never-married  Not-in-family   
9    42     Bachelors     Married-civ-spouse        Husband   
10   37  Some-college     Married-civ-spouse        Husband   
11   30     Bachelors     Married-civ-spouse        Husband   
12   23     Bachelors          Never-married      Own-child   
13   32    Assoc-acdm          Never-married  Not-in-family   
15   34       7th-8th     Married-civ-spouse        Husband   
16   25       HS-grad          Never-married      Own-child   
17   32       HS-grad          Never-married      Unmarried   
18   38          11th     Married-civ-spouse        Husband   
19   43       Masters               Divorced      Unmarried   
20   40     Doctorate     Married-civ-spouse        Husband   

                  race     sex  hoursperweek  nativecountry  income  
0                White    Male            40  United-States       0  
1                White    Male            13  United-States       0  
2                White    Male            40  United-States       0  
3                Black    Male            40  United-States       0  
4                Black  Female            40           Cuba       0  
5                White  Female            40  United-States       0  
6                Black  Female            16        Jamaica       0  
7                White    Male            45  United-States       1  
8                White  Female            50  United-States       1  
9                White    Male            40  United-States       1  
10               Black    Male            80  United-States       1  
11  Asian-Pac-Islander    Male            40          India       1  
12               White  Female            30  United-States       0  
13               Black    Male            50  United-States       0  
15  Amer-Indian-Eskimo    Male            45         Mexico       0  
16               White    Male            35  United-States       0  
17               White    Male            40  United-States       0  
18               White    Male            50  United-States       0  
19               White  Female            45  United-States       1  
20               White    Male            60  United-States       1

Elabora una regla para predecir la variable `income`, y almacena los resultados en la variable `income_pred`. Puedes comprobar cómo de buena es tu regla utilizando la función acierto, que recibe dos vectores, `income` que es el valor real e `income_pred` que es la predicción.

In [0]:
def acierto(income, income_pred):
    ac = sum(income == income_pred)*100/(len(income))
    return "Porcentaje de acierto: " + str(ac) + "%"


Este es un ejemplo muy sencillo. La regla es: diremos que cobran más de 50k aquelos cuyo nivel de estudios sea  doctorado

In [0]:
inc["income_pred"] = 0

condicion = (inc.education.isin(["Doctorate"])) 
inc.loc[condicion, "income_pred"] = 1
acierto(inc.income, inc.income_pred)

'Porcentaje de acierto: 76.53699418350115%'

Intenta buscar una regla que supere el 80% de cierto